### DISCUSSION TITLE
rewarding or penalizing constraint matches

### DISCUSSION THREAD

**Discussion by pilioulian** [TimefoldAI/timefold-solver#185] [General]
Hi,

I was wondering about whether there is a recommended way of weighing constraints. Often constraints can either be modeled by penalizing as well as rewarding something, yielding the same result (e.g. to schedule something as early as possible penalize by begin or reward by span-begin). 
Is there a best practice or does it even affect the solver when combining multiple constraints using a mix of penalize and reward?
From my understanding it should not matter, except maybe making it harder to understand the implications of single constraints in relation to the rest of the constraints

**Comment by triceo** [comment]
To the solver, the only thing that matters is the final score. Scores are ordered from best to worse, higher number means better score - whether you arrived there through rewards or penalties is of little consequence.

Even though there may not be difference between how the two are handled, there is a difference in how you think about them. Some constraints may be naturally rewarding, some naturally penalizing. If the solver only offerred penalizing, for example, that may be too limiting for constraint design. But again, practically speaking, the solver won't mind - it only cares for the final score.
